src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -74,25 +74,37 @@
 
 use crate::fx::{FxHashMap, FxHashSet};
 
-use std::cell::Cell;
+use std::cell::{Cell, RefCell};
+use std::cmp::Ordering;
 use std::collections::hash_map::Entry;
+use std::collections::BinaryHeap;
 use std::fmt::Debug;
 use std::hash;
 use std::marker::PhantomData;
+use std::mem;
 
 mod graphviz;
 
 #[cfg(test)]
 mod tests;
 
 pub trait ForestObligation: Clone + Debug {
+    /// A key used to avoid evaluating the same obligation twice
     type CacheKey: Clone + hash::Hash + Eq + Debug;
+    /// The variable type used in the obligation when it could not yet be fulfilled
+    type Variable: Clone + hash::Hash + Eq + Debug;
+    /// A type which tracks which variables has been unified
+    type WatcherOffset;
 
     /// Converts this `ForestObligation` suitable for use as a cache key.
     /// If two distinct `ForestObligations`s return the same cache key,
     /// then it must be sound to use the result of processing one obligation
     /// (e.g. success for error) for the other obligation
     fn as_cache_key(&self) -> Self::CacheKey;
+
+    /// Returns which variables this obligation is currently stalled on. If the slice is empty then
+    /// the variables stalled on are unknown.
```

Comment:
I agree that it is `None == unknown` is clearer, but an empty slice would still have to be treated as unknown, or it would need to panic on any empty slice. Otherwise it would stall forever.